edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ClientTokenCloudConnection.cs

```diff
@@ -0,0 +1,331 @@
+                if (IsTokenUsable(this.Identity.IotHubHostName, token))
+                {
+                    if (retrying)
+                    {
+                        Events.NewTokenObtained(this.Identity, token);
+                    }
+                    else
+                    {
+                        Events.UsingExistingToken(this.Identity.Id);
+                    }
+
+                    return token;
+                }
+                else
+                {
+                    Events.TokenNotUsable(this.Identity, token);
+                }
+
+                bool newTokenGetterCreated = false;
+                // No need to lock here as the lock is being held by the refresher.
+                TaskCompletionSource<string> tcs = this.tokenGetter
+                    .GetOrElse(
+                        () =>
+                        {
+                            Events.SafeCreateNewToken(this.Identity.Id);
+                            var taskCompletionSource = new TaskCompletionSource<string>();
+                            this.tokenGetter = Option.Some(taskCompletionSource);
+                            newTokenGetterCreated = true;
+                            return taskCompletionSource;
+                        });
+
+                // If a new tokenGetter was created, then invoke the connection status changed handler
+                if (newTokenGetterCreated)
+                {
+                    // If retrying, wait for some time.
+                    if (retrying)
+                    {
+                        await Task.Delay(TokenRetryWaitTime);
+                    }
+
+                    this.ConnectionStatusChangedHandler(this.Identity.Id, CloudConnectionStatus.TokenNearExpiry);
+                }
+
+                retrying = true;
+                // this.tokenGetter will be reset when this task returns.
+                token = await tcs.Task;
+            }
+        }
+
+        // Checks if the token expires too soon
+        static bool IsTokenUsable(string hostname, string token)
+        {
+            try
+            {
+                return TokenHelper.GetTokenExpiryTimeRemaining(hostname, token) > TokenExpiryBuffer;
+            }
+            catch (Exception e)
+            {
+                Events.ErrorCheckingTokenUsable(e);
+                return false;
+            }
+        }
+
+        class ClientTokenBasedTokenProvider : ITokenProvider
+        {
+            readonly ClientTokenCloudConnection cloudConnection;
+            string token;
+
+            public ClientTokenBasedTokenProvider(ITokenCredentials tokenCredentials, ClientTokenCloudConnection cloudConnection)
+            {
+                this.cloudConnection = cloudConnection;
+                this.token = tokenCredentials.Token;
+            }
+
+            public async Task<string> GetTokenAsync(Option<TimeSpan> ttl)
+            {
+                using (await this.cloudConnection.tokenUpdateLock.LockAsync())
```

Comment:
Can this lock be moved to ClientTokenBasedTokenProvider?

Reply:
Yep, will move.

Reply:
Moved it.. but I wonder if it was in the parent for a reason.. (synchronizing multiple token callbacks..).. Let me think about it a bit more.